### ISSUE TITLE
[freetype] freetype should accept libpng or libpng-apng as requirement

### ISSUE THREAD

**Issue by leleco95** [microsoft/vcpkg#8599]
****

**Is your feature request related to a problem? Please describe.**
I installed libpng-apng and then went to install freetype, but it requires libpng.
The current way it works is libpng-apng is a patch that does not have a dependency to libpng, so I "don't have" libpng installed. Therefore, both libs/dlls are the same and vcpkg identified that the files already existed and aborted install.

```
The following packages will be built and installed:
    freetype[core]:x64-windows
  * libpng[core]:x64-windows
Additional packages (*) will be modified to complete this operation.
Starting package 1/2: libpng:x64-windows
Building package libpng[core]:x64-windows...

-- Using cached C:/GitHub/vcpkg/downloads/glennrp-libpng-v1.6.37.tar.gz
-- Using source at C:/GitHub/vcpkg/buildtrees/libpng/src/v1.6.37-6445e5863e
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Performing post-build validation
-- Performing post-build validation done
Building package libpng[core]:x64-windows... done
Installing package libpng[core]:x64-windows...
The following files are already installed in C:/github/vcpkg/installed/x64-windows and are in conflict with libpng:x64-windows

Installed by libpng-apng:x64-windows
    bin/libpng16.dll
    bin/libpng16.pdb
    debug/bin/libpng16d.dll
    debug/bin/libpng16d.pdb
    debug/lib/libpng16d.lib
    include/libpng16/png.h
    include/libpng16/pngconf.h
    include/libpng16/pnglibconf.h
    include/png.h
    include/pngconf.h
    include/pnglibconf.h
    lib/libpng16.lib
```

**Proposed solution**
I have two proposed solutions:
1. Make it so libpng-apng installs libpng as a dependency
2. Make it so freetype accepts either libpng or libpng-apng (whichever is present)

The problem I see for the first one would be version inconsistency, such as libpng releases a new version but libpng-apng doesn't, so it's not possible to update one without the other.
Meanwhile, the second solution isn't exactly accurate, because having libpng-apng installed should count as having libpng installed. In my case, I solved it by changing the CONTROL file from freetype so it would require libpng-apng instead of libpng, but it should not be the ideal solution.

**Comment by Neumann-A** [comment]
hmm so we need a png metapacket which the other ports depend on using features in the metaport to differentiate between the png libraries.

**Comment by AlvinZhangH** [comment]
Hi @leleco95 thanks for posting this issue. This is a known vcpkg problem. Port libpng-apng and libpng have the same home page, when one of them is installed, delete the other.

**Comment by leleco95** [comment]
Hey @AlvinZhangH, thanks for your attention.

In my project, I need freetype and libpng-apng installed, but freetype makes me install libpng because it is a dependency and, if I try to uninstall libpng after that (so I can install libpng-apng), it tells me freetype must be uninstalled also, so with the current setup I'd be stuck in a loop of:

- if I have libpng-apng, I can't install freetype
- if I have freetype, I can't install libpng-apng

As I mentioned earlier, I solved my case by modifying freetype's CONTROL file so it would take libpng-apng as a dependency instead of libpng, but it's not a solution that should be implemented, since someone else could simply want freetype and libpng.

Since you said this is a known problem, is it in the sense that I should stick to my workaround or that it might be solved some time in the future?

Thanks in advance.

**Comment by vicroms** [comment]
> Since you said this is a known problem, is it in the sense that I should stick to my workaround or that it might be solved some time in the future?

This is the an issue we have faced with other libraries that are mutually exclusive, e.g.: `openssl` vs `libressl`.

Our current strategy has been to have one of them become the "blessed" package and have all other libraries that need an implementation depend on the "blessed" library. This also helps prevent file conflict errors during CI (where we build every package at the same time, and if we try to build conflicting packages the build will fail every time).

We decide if a meta-package is worth to implement on a case-by-case basis, such as the `opencv` meta-package.

**Comment by JackBoosY** [comment]
Hi @leleco95, I create a PR #8622 to fix this issue, can you modify the freetype dependency from `libpng` to `libpng[apng]` and rebuild freetype?

Thanks.

**Comment by JackBoosY** [comment]
@leleco95 So, I think PR #8622 fixed this issue, I will close this issue after #8622 merge.

**Comment by leleco95** [comment]
Yes and it will also fix #8598. Thank you very much for your help.

**Comment by JackBoosY** [comment]
Close this issue via #8622 merged.
